subsys: Add MCTP as a subsystem, built on libmctp with bindings

teburd opened the pull request:
Adds libmctp as a zephyr module from my fork with a few small changes I intend on trying to upstream along with a binding to Zephyr's ~polling~ async UART API. Comes with a pair of sample applications that can be used on two boards connected via UART.

Commit:
modules: Add libmctp as a module

Adds libmctp as a west module dependency

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

KeHIntel commented on MAINTAINERS.yml:
```diff
@@ -4743,6 +4743,17 @@ West:
+  maintainers:
+    - teburd
+  collaborators:
+    - nashif
```
Hi, Tom

Can you add @inteljiangwe1 who as collaborator?
Wei is our owner on MCTP, who can help work on this module, and add other components.

BR,
Ke Han

Commit:
MAINTAINERS.yml: Add libmctp module maintainership

Adds myself, nashif, and inteljiangwe1 to the maintainers file covering
the libmctp module.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
samples: Add mctp host and client samples

Samples work by sending MCTP encoded messages over a uart between two
boards.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Need to work out some upstream stuff, but can then have the fork setup in the zephyrproject-rtos github org

Commit:
mctp: Add mctp subsystem with uart binding

libmctp provides interfaces for wiring up a MCTP bus it calls bus
bindings. The bindings provided in libmctp however are not directly
useful to Zephyr without some work. Provide an initial uart binding that
directly uses Zephyr's async uart interface.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
In my original samples I had hacked up the serial binding to work with Zephyr natively, to better integrate with Zephyr however its useful to provide native bindings in a subsys so that's what I've begun to do.

henrikbrixandersen commented on doc/Makefile:
```diff
@@ -4,7 +4,7 @@
 BUILDDIR ?= _build
 DOC_TAG ?= development
-SPHINXOPTS ?= -j auto -W --keep-going -T
+SPHINXOPTS ?= -j2 -W --keep-going -T
```
Stray change?

teburd replied:
Fixed

henrikbrixandersen commented on samples/modules/mctp/mctp_endpoint/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(i2c_peripheral)
```
Wrong name?

teburd replied:
Fixed this and the mctp host project name as well (was also wrong)

KeHIntel commented:
Add @inteljiangwe1 to help review this PR

teburd commented:
@henrikbrixandersen hopefully this shows better why Zephyr integration is needed with the latest binding and samples. I'd also note several folks are interested in this working with Zephyr including @KeHIntel @dkalowsk and @beriberikix

There's still more I can envision happening here as well, multiplexing things over say the console uart (or any other transport) with mctp would be really nice (e.g. endpoints for shell/logging/tracing/etc). And is something I want to look into.

dkalowsk commented:
Just confirming that the module repo would be moved to the zephyrproject and not continue to be housed in your personal GitHub account.

teburd commented:
> Just confirming that the module repo would be moved to the zephyrproject and not continue to be housed in your personal GitHub account.

Yes it would be, and my changes upstreamed before merging. DNM until this happens.

I considered how I might write an MCTP uart test but this becomes pretty annoying to deal with. It's probably easier to have the samples become tests with a fixture (board to board) somehow. I'm very open to suggestion on the testing side of this, the samples absolutely work well.

KeHIntel commented:
> @henrikbrixandersen hopefully this shows better why Zephyr integration is needed with the latest binding and samples. I'd also note several folks are interested in this working with Zephyr including @KeHIntel @dkalowsk and @beriberikix
> 
> MCTP is also used in Meta's OpenBIC (https://github.com/facebook/OpenBIC) with Zephyr as well, though with its own implementation which... is exactly what this would help against.
> 
> There's still more I can envision happening here as well, multiplexing things over say the console uart (or any other transport) with mctp would be really nice (e.g. endpoints for shell/logging/tracing/etc). And is something I want to look into. This might require some DT integration to point logging/console at a pseudo uart (mctp endpoint). So perhaps describing mctp endpoints in DT would become a thing then.

KeHIntel commented:
Sorry, mistakenly clicked the close button 😂
Just want to see MCTP binding with I3C code, either in this PR or another PR. Thanks.

teburd commented:
> Sorry, mistakenly clicked the close button 😂 Just want to see MCTP binding with I3C code, either in this PR or another PR. Thanks.

I3C binding indeed is on my todo list! :+1:

teburd commented:
Small fixes to comments/board overlays since the initial PR passed CI and you +1'ed @dkalowsk, thanks!

kartben reviewed (changes requested):
May I please request a README for the code sample? I appreciate the effort on the Doxygen docs—I really do!— but I think this needs slightly more, and the readme will be a great way to make this new subsystem more visible (and it is strictly speaking a requirement for all code samples anyway ;))

teburd commented:
> May I please request a README for the code sample? I appreciate the effort on the Doxygen docs—I really do!— but I think this needs slightly more, and the readme will be a great way to make this new subsystem more visible (and it is strictly speaking a requirement for all code samples anyway ;))

Done

henrikbrixandersen commented on samples/modules/mctp/mctp_endpoint/src/main.c:
```diff
@@ -0,0 +1,68 @@
+{
+	switch (eid) {
+	case REMOTE_HELLO_EID:
+		LOG_INF("got mctp message %s for eid %d, replying to 5 with \"world\"", (char *)msg,
```
What does "replying to 5 with" mean here?

teburd replied:
The endpoint it’s assuming it’s replying to, I can adjust the message string here

teburd replied:
The endpoint it’s replying to, though I get that’s not clear, and not even known so I’ll correct this

dkalowsk commented on subsys/mctp/mctp_uart.c:
```diff
@@ -0,0 +1,331 @@
+	for (i = 0, j = 0; i < total_len; i++, j++) {
+		uint8_t c = p[i];
+
+		if (c == 0x7e || c == 0x7d) {
```
minor nit: instead of the magic sentinel numbers here, would it make sense to use the values from lines 18 and 19?  Asking as my own dislike is to magic numbers and needing to figure out what is going on.  They might need a different name to ensure there is no name collision with MCTP's defines.

dkalowsk commented on subsys/mctp/mctp_uart.c:
```diff
@@ -0,0 +1,331 @@
+{
+	struct mctp_pktbuf *pkt = uart->rx_pkt;
+
+	__ASSERT_NO_MSG(pkt);
```
suggestion: move this ASSERT into the `if (valid)` statement.  No need to shutdown the whole system if it was already going to be a no-op anyways.

dkalowsk commented on subsys/mctp/mctp_uart.c:
```diff
@@ -0,0 +1,331 @@
+
+static void mctp_uart_start_pkt(struct mctp_binding_uart *uart, uint8_t len)
+{
+	uart->rx_pkt = mctp_pktbuf_alloc(&uart->binding, len);
```
A common practice I see often is developers re-using variables.  In this case would it make sense to ASSERT on entering if the `uart->rx_pkt` is not NULL?  I believe the whole `mctp_pktbuf_alloc` does actually do an `m_alloc` or `m_realloc` which could result in a memory leak here.

teburd replied:
Yes it should be NULL when allocating, https://github.com/zephyrproject-rtos/zephyr/pull/75743/files#diff-c9d952605a6e4b48ea6b1e57729eb4bd6ab93f3f3890976b73b04fe43ac3ac60R59 NULLs the pointer once its moved into libmctp's call (which eventually I believe leads to the application callbacks and such).

dkalowsk reviewed (dismissed):
I've got a few minor suggestions, but nothing that should stop this from moving forward.  Approving on that end.

kartben commented on samples/modules/mctp/mctp_endpoint/README.md:
We do .rst :) also please follow the code sample template https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/templates/sample.tmpl so this properly shows in the documentation. You'll also need a page similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/modules/lvgl/lvgl.rst in the mctp folder to define a new sample category for MCTP samples

teburd replied:
The sample here is really the pair of firmwares being built and run on two boards. Do we have anything like that in the tree already that you know of?

Really this is one sample with two different firmwares. Really unsure of how best to structure it but documenting them separately actually doesn't make much sense.

teburd replied:
I think its good now, let me know

dkalowsk commented:
seems to have picked up an extra commit for renaming `llext_simple_test` and not sure if that is correct.  The MCTP changes all look good to me.  Approval held off due to confusion on the extra commit.

teburd commented:
> seems to have picked up an extra commit for renaming `llext_simple_test` and not sure if that is correct. The MCTP changes all look good to me. Approval held off due to confusion on the extra commit.

Oops! yeah had committed that to my local main, and now all my rebases have it, argh, gotta go review other PRs as well.

teburd commented:
Now passing CI

kartben commented on samples/modules/mctp/mctp_endpoint/README.rst:
```diff
@@ -0,0 +1,43 @@
+.. zephyr:code-sample:: mctp_endpoint_sample
+   :name: MCTP Endpoint Sample
+
+   This sample shows how to create an MCTP endpoint (a reactive node in a
+   network of MCTP nodes) over UART. MCTP is a multiplexed messaging protocol
+   that works over a variety of transports (i2c, i3c, uart, pcie, etc).
```
This is meant to be the (very) short description that shows up in the code sample catalog (https://docs.zephyrproject.org/latest/samples/index.html). Suggested rewording as per code sample README template below (for example, it shouldn't be starting with "This sample shows...", but you may want to move some of the original text to the Overview section as it probably contains useful bits

```suggestion
   Create an MCTP endpoint over UART.
```

kartben commented on samples/modules/mctp/mctp_host/README.rst:
```diff
@@ -0,0 +1,43 @@
+.. zephyr:code-sample:: mctp_host_sample
+   :name: MCTP Host Sample
+
+   This sample shows how to create an MCTP host (a requesting node in a network
+   of MCTP nodes) over UART. MCTP is a multiplexed messaging protocol that works
+   over a variety of transports (i2c, i3c, uart, pcie, etc).
```
```suggestion
   Create an MCTP host over UART.
```

kartben commented on samples/modules/mctp/mctp_endpoint/README.rst:
```diff
@@ -0,0 +1,43 @@
+Overview
+********
+Sets up an MCTP node that listens on a UART for messages targeting a particular
+MCTP endpiont id with the message "hello". Responds to this "hello" message with
```
```suggestion
MCTP endpoint id with the message "hello". Responds to this "hello" message with
```

kartben commented on samples/modules/mctp/mctp_endpoint/README.rst:
```diff
@@ -0,0 +1,43 @@
+Wiring
+******
+The listening UART pins should be wired to a board which will run the MCTP host
+sample such that this boards UART tx pin connects to the host boards rx pin,
+and this boards UART rx pin connects to the host boards tx pin. The boards
+grounds should also be wired together.
```
```suggestion
sample such that this board's UART tx pin connects to the host boards rx pin,
and this board's UART rx pin connects to the host board's tx pin. The boards'
grounds should also be wired together.
```

kartben commented on samples/modules/mctp/mctp_host/README.rst:
```diff
@@ -0,0 +1,43 @@
+Wiring
+******
+The UART pins should be wired to a board which will run the MCTP endpoint
+sample such that this boards UART tx pin connects to the endpoint boards rx
+pin, and this boards UART rx pin connects to the endpoint boards tx pin. The
+boards grounds should also be wired together.
```
```suggestion
sample such that this board's UART tx pin connects to the endpoint board's rx
pin, and this board's UART rx pin connects to the endpoint boards tx pin. The
boards' grounds should also be wired together.
```

kartben commented on subsys/mctp/mctp_uart.c:
```diff
@@ -0,0 +1,325 @@
+			 * and the state machine is out of sync. The failed request's
+			 * trailing footer i.e. 0x7e would be interpreted as next
+			 * request's framing footer. So if we are in STATE_WAIT_REVISION
+			 * and receive 0x7e byte, then contine to stay in
```
nit 
```suggestion
			 * and receive 0x7e byte, then continue to stay in
```

kartben commented on samples/modules/mctp/mctp_host/README.rst:
```diff
@@ -0,0 +1,43 @@
+
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/modules/mctp/mctp_endpoint
```
```suggestion
   :zephyr-app: samples/modules/mctp/mctp_host
```

ofirshe commented:
Hi,
Upon rebasing to the latest Zephyr after long time, I’m seeing this warning:
```
CMake Warning at zephyr/CMakeLists.txt:1002 (message):  
  No SOURCES given to Zephyr library: modules_mctp  
  Excluding target from build.
```
I also noticed that `CONFIG_MCTP` isn’t set in output `.config` file. If it’s disabled, why does this warning appear?
Any guidance would be appreciated!

ubieda commented:
> Hi, Upon rebasing to the latest Zephyr after long time, I’m seeing this warning:
> 
> ```
> CMake Warning at zephyr/CMakeLists.txt:1002 (message):  
>   No SOURCES given to Zephyr library: modules_mctp  
>   Excluding target from build.
> ```
> 
> I also noticed that `CONFIG_MCTP` isn’t set in output `.config` file. If it’s disabled, why does this warning appear? Any guidance would be appreciated!

Could you try `west update` and see if it's gone?

ofirshe commented:
@ubieda I ran `west update` immediately after rebasing.
